31 implement family addition import

nilsoberg opened the pull request:
In order to fully close this issue changes were made that close issues #31 , #29 , and #147 .

nilsoberg committed:
Create robust metadata modules

- Reader for parsing, with an API rather than raw hash ref access
- Writer for saving, with an API for saving rather than raw hash ref access

nilsoberg committed:
Refactor EST import config for clarity

- Code reuse by inheriting from EFI::Options
- Pass necessary hash data to EFI::Import::Source modules rather than Config object
- Removed a lot of unnecessary code in the Config modules

nilsoberg committed:
Add sequence collection module for a single file to replace accession and metadata files

nilsoberg committed:
Centralize output directory argument retrieval for import utilities

nilsoberg committed:
Move filter code into separate script

nilsoberg committed:
Refactor sequence ID retrieval for modularity and implement family addition

- Remove sunburst and stats code
- Remove filter-specific code (moved in previous commit)
- Metadata is managed by EFI::Sequence::Collection and not individual Sources
- Sequences are added to a sequence collection by the source, (i.e. the source
  no longer returns a sequence collection), which allows family addition

nilsoberg committed:
Logical command line arguments indicating the step in pipeline

- get_sequence_ids outputs 'source' files -> filter_ids
- filter_ids outputs main sequence_metadata and accession_ids files
- Add central sequence typing
- Provide extra error message output to printHelp in EFI::Options
- Config subclasses can update options for use by calling scripts

nilsoberg committed:
Proper UniRef support in EST import process

- get_sequence_ids outputs a metadata file and a UniRef mapping table (used to be accession_ids.txt)
- Filtering removes from metadata as well as UniRef mapping
- Support dual file approach and UniRef in EFI::Sequence::Collection
- Update filter_ids filter order

nilsoberg committed:
Filter using new EFI::Sequence::Collection dual file approach

- Fix filter bugs
- Fix EFI::Sequence api bug

nilsoberg committed:
Add sunburst json file output scripts

nilsoberg committed:
Add predefined taxonomy filters

nilsoberg committed:
Make UniRef sequence collection API consistent

nilsoberg committed:
Move sequence attribute formatting to EFI::Sequence

nilsoberg committed:
Support removal of members of UniRef clusters from a sequence collection

nilsoberg committed:
Share sequence filtering code between modules

nilsoberg committed:
Add UniRef support to filtering and sequence collection

nilsoberg committed:
Refactor Nextflow EST pipeline to support UniRef, metadata, and filtering

- Filtering moved to separate process
- UniRef moved to get_sequences process
- Metadata retrieval occurs in filtering
- Add sunburst ID retrieval

nilsoberg committed:
Add centralized unknown ID detection capability

nilsoberg committed:
Add unmatched IDs as output from sequence ID retrieval process

nilsoberg committed:
Centralize database schema constants

nilsoberg committed:
Validate UniProt IDs by default when reverse mapping IDs

nilsoberg committed:
Support parsing option groups

nilsoberg committed:
Add documentation and minor code cleanup

nilsoberg committed:
Centralize batch retrieval by using grouped conditions in SQL

nilsoberg committed:
Name test module output directories based on the script file name

nilsoberg committed:
Add tests for new import features and combinations

nilsoberg committed:
Support family addition to FASTA import option

- Previous accession_ids file is now accession_table to better reflect purpose
- Output of step 1/step 2 is sequence IDs file (to replace accession IDs), which is
  used in the split/get_sequence processes
- import_fasta outputs a file that contains only sequence IDs that were added from
  families - this gets passed to split/get processes
- Support empty files in split/get_sequence processes

nilsoberg committed:
Support filter stats in import stats file output

- Statistics can now read/write
- Source outputs a stats file, filter modifies it to the expected import_stats.json

nilsoberg committed:
Cleanup to remove unnecessary code

nilsoberg committed:
Remove family source import option and update tests to reflect that change

nilsoberg committed:
BLAST database import version is no longer tied to use of UniRef

nilsoberg committed:
Improve UniRef support

- Fix issue with retrieving UniRef sequences for Options A, C, and D
- Properly retrieve UniRef IDs for Option B

nilsoberg committed:
Implement proper UniRef support in sequence collection and fix removal bugs

nilsoberg committed:
Optimize taxonomy data retrieval for sunburst

nilsoberg committed:
Fix wrong test parameters and confusing test naming

nilsoberg committed:
Fix issues with UniRef metadata retrieval and filtering

nilsoberg committed:
Fix wrong BLAST test parameters and UniRef test naming

nilsoberg committed:
Fix test case

nilsoberg committed:
Remove unused code and add documentation

nilsoberg committed:
Create documentation for EFI::Sequence::Collection module

rbdavid commented on tests/test_env.sh:
```diff
@@ -79,6 +80,9 @@ do
 	elif [[ ${!index} == "--fasta-db" ]]; then
 		fasta_db="${!idx}"
 		echo "Using $fasta_db as the FASTA database path"
+	elif [[ ${!index} == "--blast-importfasta-db" ]]; then
+		blast_import_fasta_db="${!idx}"
+		echo "Using $blast_import_fasta_db as the UniRef BLAST import FASTA database path"
```
Add a doc help string for this input parameter to lines 40-58.

rbdavid replied:
It may be semi-redundant with the `--fasta-db` input arg's help string but the differences are important so should be explicitly described.

rbdavid commented on tests/test_env.sh:
```diff
@@ -129,12 +133,16 @@ fi
 if [[ -z "$fasta_db" ]]; then
     fasta_db="$DATA_DIR/blastdb/combined.fasta"
 fi
+if [[ -z "$blast_import_fasta_db" ]]; then
```
To keep in formatting, include a white space line btwn these if statements.

rbdavid commented on pipelines/est/est.nf:
```diff
@@ -1,53 +1,96 @@
+        path 'accession_table.tab', emit: 'accession_table'     // table of all sequence IDs, including UniRef IDs, filtered
+        path 'sequence_metadata.tab', emit: 'sequence_metadata' // sequence metdata in metadata format
+        path 'sequence_ids.tab', emit: 'sequence_ids'           // list of primary sequence IDs in the metadata file
+        path 'import_stats.json', emit: 'import_stats'          // statistics of source and filter import processes
```
This file being emitted as `import_stats` but goes unused in the workflow block. Can remove the `emit: 'import_stats'` portion since the file is listed in output purely so that it gets copied to the `publishDir`.

nilsoberg replied:
Removed emit

rbdavid commented on pipelines/est/est.nf:
```diff
@@ -1,53 +1,96 @@
-    cp $existing_fasta_file allsequences.fa
-    """
+        path 'seq_mapping.tab', emit: 'seq_mapping', optional: true
+        path 'unmatched_id.tab', emit: 'unmatched_ids', optional: true
```
This file is being emitted as `unmatched_ids` but goes unused in the workflow block. Can remove the `emit: 'unmatched_ids'` portion since the file is listed in output purely so that it gets copied to the `publishDir`.

rbdavid commented on pipelines/est/est.nf:
```diff
@@ -1,53 +1,96 @@
+        path 'source_ids.tab', emit: 'source_ids'
+        path 'source_seq.tab', emit: 'source_meta'
+        path 'source_stats.json', emit: 'source_stats'
         path 'blast_hits.tab', emit: 'blast_hits', optional: true
```
This file is being emitted as `blast_hits` but goes unused in the workflow block. Is this implemented for future use?

rbdavid commented on pipelines/est/est.nf:
```diff
@@ -188,46 +228,66 @@ process visualize {
+    sequence_id_files = filter_ids(source_data.source_ids, source_data.source_meta, source_data.source_stats)
+
+    // Get sunburst data for all sequence IDs, after filtering
+    get_sunburst_data(sequence_id_files.accession_table, sequence_id_files.sequence_metadata)
```
The `get_sunburst_data` process emits one file path as `sunburst_tax` but, since output from this call of the process is not saved to a variable name space, that emitted file path cannot be accessed/used. Luckily, it doesn't get used anywhere below, but maybe it will? Should we save the output from this process into a variable?

rbdavid replied:
Or, as suggested in the comment for the `get_sunburst_data` process, the emit portion of that output block could be removed.

nilsoberg replied:
It's probably not necessary in future processes, however I will remove the emit portion.

rbdavid commented on pipelines/est/est.nf:
```diff
@@ -1,53 +1,96 @@
+        path accession_table
+        path sequence_metadata
+    output:
+        path 'sunburst_tax.json', emit: 'sunburst_tax'
```
This file is emitted as `sunburst_tax` but goes unused in the workflow block. Can remove the `emit: 'sunburst_tax'` portion since the file is listed in output purely so that it gets copied to the `publishDir`.

rbdavid commented on pipelines/est/est.nf:
```diff
@@ -188,46 +228,66 @@ process visualize {
     }
 
-    // step 2: multiplex
+    // Add the imported FASTA file if the import mode is fasta
```
This comment goes with the above `if` statement.

rbdavid commented on pipelines/est/est.nf:
```diff
@@ -188,46 +228,66 @@ process visualize {
+    // from the BLAST sequence database.  If the import mode is FASTA, then these IDs are only
+    // ones that come from adding a family to the job
+    accession_shards = split_sequence_ids(sequence_ids, params.num_accession_shards)
+    fasta_files = get_sequences(accession_shards.flatten(), params.fasta_db)
```
The `get_sequences` process is run `params.num_accession_shards` times; this is done to enable parallelization of this step, but none of nextflow config files have this process listed with non-standard executor/cpus/memory/etc configurations, so its being done serially AFAIK.

nilsoberg replied:
It's set as a default command line argument in `bin/create_est_nextflow_params.py` (line 21).  Is that what you were looking for?

rbdavid replied:
My confusion came from not recognizing that accession_shards became a list of files. That list of files is handed to the `get_sequences` processes, which is run in parallel because Nextflow is smart. For each of the `get_sequences` tasks, nextflow used the default process setup defined in the config file (e.g. `conf/est/docker.config`). Anyways, this works great.

rbdavid commented on bin/create_est_nextflow_params.py:
```diff
@@ -130,21 +137,15 @@ def render_params(output_dir, duckdb_memory_limit, duckdb_threads, fasta_shards,
         "efi_db": efi_db,
         "import_mode": import_mode,
-        "exclude_fragments": exclude_fragments,
+        "filter": sequence_filter,
```
Could the variable/parameter name be more consistent throughout this script. It changes from the input argument `--filter` to `args.sequence_filter` to the `sequence_filter` input argument for the `render_params` function, and finally, to the `"filter"` key in the params dict that is written out to the json file.

nilsoberg replied:
This is an unfortunate necessity.  I'd rather keep `--filter` as the entry point and the key in the yml because that's the same in the perl scripts, and since it can be used multiple times it's shorter than `--sequence-filter`.  Python has a built-in function called `filter`, so when you expand the namespace into arguments for `render_params` you can't use `filter` as a variable name.  It gets renamed at line 101.  Does that clarify?

rbdavid replied:
Variable namespaces used in a function are local to that function. So, using `filter` as an input argument name in `render_params` will overwrite the built-in python function `filter()`, sure. But, outside of that function, `filter` will still map to the namespace for the builtin filter function. It may be bad practice to use the builtin functions'/objects/etc namespaces, but we could do it if we wanted to. 

An alternative is, on line 27 in `add_args`, we could implement multiple acceptable command line argument names that all map to one argparse.Namespace attribute. 

```python3
common_parser.add_argument("--sequence-filter", "--filter", action="append", type=str, help="Filter sequences, use multiple times to indicate filter types")
# this enables either --sequence-filter _or_ --filter to be used on the command line to signify 
# this input argument. Since --sequence-filter is defined first in the .add_argument() call
# the argparse.Namespace will have the filter values stored in args.sequence_filter
``` 

This might be the cleanest way to avoid jumping across various names while still letting the command line argument be input via `--filter` and written out to json as "filter". Low priority/importance.

rbdavid commented on lib/EFI/Sequence.pm:
```diff
@@ -0,0 +1,310 @@
+
+
+# public
+sub getAttribute {
+    my $self = shift;
+    my $attr = shift || die "Require attribute name";
+    my $val = $self->{attr}->{$attr};
+    return $val;
+}
```
How does this handle an attribute name that has not been set for the Sequence object/instance?

rbdavid replied:
Should it throw a fatal error or an empty string?

nilsoberg replied:
If no attribute name is provided (e.g. the function is called without any arguments or with an undefined or empty name) then the function currently dies.  (e.g. throws a fatal error.)

rbdavid commented on lib/EFI/Sequence.pm:
```diff
@@ -0,0 +1,310 @@
+    # If multiple values were passed, then convert to an array ref
+    if (not ref $vals[0] and @vals > 1) {
+        $val = \@vals;
+    } elsif (ref $vals[0]) {
+        $val = $vals[0];
+    } else {
+        $val = $vals[0];
+    }
```
Okay, let me test my perl knowledge. The code is checking if the `ref $vals[0]` is actually a reference to some value, returning an empty string if not. So, perl has a similar truthiness expectation that empty strings are equivalent to `False`. Great! 

But the code here is doing the exact same thing, whether the `ref $vals[0]` call returns an empty string or not. So why include the `elsif` block and just leave the `else` block? 

Also, https://perldoc.perl.org/functions/ref suggests that a `ref` call can return `0`, which is also equivalent to False but also a valid return result from the `ref` command. So, in this case, I think an explicit boolean check for empty strings should be implemented, e.g. `(ref $vals[0]) eq ""`.

nilsoberg replied:
The `elsif (ref` was removed -- holdover from another way of doing things.

rbdavid commented on lib/EFI/Sequence.pm:
```diff
@@ -0,0 +1,310 @@
+# public
+sub setAttribute {
+    my $self = shift;
+    my $attr = shift;
```
What formatting is acceptable for attribute names? Just any string? White spaces in it? Should any validation of the attribute name be applied here?

nilsoberg replied:
If this was more general purpose then probably better validation should happen here.  Any valid hash (dictionary) key is technically acceptable here, basically any value except `undef`.  Within the workflows, the scripts will use values from **`EFI::Annotations::Fields`** as attribute names.

rbdavid commented on lib/EFI/Sequence.pm:
```diff
@@ -0,0 +1,310 @@
+                push @vals, $part;
+            }
+        }
+        return join("^", @vals);
```
Would it be helpful to enable any delimiter character(s) to be used? Currently, if the attribute value array contains a "^" character in it, any down-stream parsing code may have issues parsing this packed string.

rbdavid replied:
Using a multi-character delimiting string could help avoid instances where a stray "^" character makes its way into protein annotations (or any other attribute fields associated with a sequence).

nilsoberg replied:
This is a good idea, and I added the functionality.

rbdavid commented on lib/EFI/Sequence.pm:
```diff
@@ -0,0 +1,310 @@
+sub unpackAttributeValue {
+    my $self = shift;
+    my $value = shift;
+    my @parts = split("^", $value);
```
See above, about enabling any delimiter to be used instead of hard-coded "^".

rbdavid commented on lib/EFI/Sequence/Collection.pm:
```diff
@@ -0,0 +1,976 @@
+# Parameters:
+#    $outputFile - path to output file
+#
+sub saveMetadataFile {
```
If `packAttributevalue` is enabled to have any delimiter be used, this subroutine will require another argument to set that value.

nilsoberg replied:
Added it as an argument to the constructor.

rbdavid commented on lib/EFI/Sequence/Collection.pm:
```diff
@@ -0,0 +1,976 @@
+=head4 Example Usage
+
+    $seq->setAttribute("custom", "value");
+    $val = $seq->getAttribute("custom");
+    # $val is "value"
+
+    $seq->setAttribute("list1", ["item 1", "item 2", "item 3"]);
+    $val = $seq->getAttribute("list1");
+    # $val is: "item 1^item 2^item 3"
+
+    $seq->setAttribute("list2", "item 1", "item 2", "item 3");
+    $val = $seq->getAttribute("list1");
+    # $val is: "item 1^item 2^item 3"
+
+
+=head3 C<packAttributeValue($value)>
+
+Packs the attribute value into a string that can be serialized and deserialized.
+Elements in packed arrays are separated by the caret character (C<^>).
+
+=head4 Parameters
+
+=over
+
+=item C<$value>
+
+Value to pack, either a scalar or an array ref.
+
+=back
+
+=head4 Returns
+
+Returns C<$value> if scalar.  Returns packed array if C<$value> is an array ref.
+
+=head4 Example Usage
+
+    $val = $seq->packAttributeValue("value");
+    # $val is "value"
+    $val = $seq->packAttributeValue(["item 1", "item 2", "item 3"]);
+    # $val is: "item 1^item 2^item 3"
```
This is documentation for the EFI::Sequence module file. Why include it here too?

nilsoberg replied:
Copy paste.  Removed.

rbdavid commented on lib/EFI/Sequence.pm:
```diff
@@ -0,0 +1,310 @@
+}
+
+
+# public
```
I think the `getAttribute` and `setAttribute` subroutines should be immediate neighbors rather than being separated by the `getAttributeNames` subroutine. Maybe `getAttributeNames` is defined first?

nilsoberg replied:
Leaving in place to avoid cluttering history

rbdavid commented on lib/EFI/Sequence.pm:
```diff
@@ -0,0 +1,310 @@
+
+=item C<$id>
+
+UniProt sequence identifier.
```
Does this need to be a UniProt Accession ID? Can it be a UniRef identifier?

nilsoberg replied:
UniRef are also UniProt IDs so this is more generally referring to any UniProt-formatted ID.  I can explain more in person.

nilsoberg replied:
typeof UniProt === UniRef, it's simply a contextual thing

rbdavid commented on lib/EFI/Sequence.pm:
```diff
@@ -0,0 +1,310 @@
+
+=item C<seq>
+
+Optional protein sequence as a string.
```
What's the use-case of a Sequence object that has no "seq" string? Why is this optional?

nilsoberg replied:
The code add sequences in the FASTA import option but no where else, and even in this case nothing happens.  However, I designed it so that in the future an `EFI::Sequence` object could be used for assisting with domains and other future issues.

rbdavid commented on lib/EFI/Sequence/Collection.pm:
```diff
@@ -0,0 +1,976 @@
+    my $attr = shift;
+    my $seq = shift;
+
+    return 0 if $self->{seq}->{$id};
```
If the ID already exists, don't overwrite it. But, what if the 2nd or latter entry of the ID has more attribute/metadata fields? Should completeness of attributes be considered before just ignoring non-unique ID entries?

rbdavid replied:
Before returning, you could loop over the `$attr` fields and check whether those already exist in `$self->{seq}->{$id}` Sequence object.

nilsoberg replied:
Yes, in theory.  I didn't think to code it because this functionality isn't needed for the scope of this PR and issue.  I can see this being useful in the future but I'd rather solve that problem then.

rbdavid replied:
Should an issue be made for this or leave it as a design idea on a post-it note?

rbdavid commented on lib/EFI/Sequence/Collection.pm:
```diff
@@ -0,0 +1,976 @@
+
+    if ($uniref50) {
+        $uniref90 = $uniprot if not $uniref90;
+        $self->{uniref50}->{$uniref50}->{$uniref90} = 1;
```
I don't think I fully understand the purpose of the uniref50, uniref90, and uniprot hashes. On this line, you are traversing the uniref50 hash, going into the `$uniref50` subhash, and then finding the `$uniref90` key and setting its value to 1. Why? 

If the collection object already has seen the combo of `$uniref50` and `$uniref90`, you aren't recording that the collection in fact contains more than one instance of that uniref ID combo (like an `+= 1` increment would). Alternatively, you aren't explicitly tracking UniProt IDs that map to that uniref50/90 ID combo (like appending the uniprot ID to a list would). 

Sorry, I may be mixing python lingo into my discussion about perl code.

nilsoberg replied:
`uniref50` is a hash that maps UniRef50 IDs to UniRef90 IDs.  I simply want a list of which UniRef90 IDs are in a given UniRef50 cluster.  Since the (UniRef50, UniRef90) pairing may occur multiple times a hash of hashes is best.  I assign `1` to indicate that the the (UniRef50, UniRef90) pair exists.  I could increment but assigning `1` has been my habit and didn't think to increment.  I don't see an immediate benefit to it but it's easy to add.

rbdavid replied:
I guess this implementation works since we don't really care how many instances of UniRef90/UniProt Ids have been seen in the input sequences. We just use entries in the `$self->{uniref50}` or `$self->{uniref90}` hashes to denote that at least one has been seen. Then, if filtering says to remove entries associated with the UniRef50/90 Ids, we traverse their hashes to find the UniRef90/UniProt Ids that need to be removed from the respective hashes. So, ignore my suggestion about incrementing or gathering a list of IDs.

rbdavid commented on lib/EFI/Sequence/Collection.pm:
```diff
@@ -0,0 +1,976 @@
+        $self->{uniref50}->{$uniref50}->{$uniref90} = 1;
+    }
+    if ($uniref90) {
+        $self->{uniref90}->{$uniref90}->{$uniprot} = 1;
```
I have a similar confusion as above. Traversing the `uniref90` hash to get the `$uniref90` subhash, getting the `$uniprot` key and setting its value to 1. At least here, the `$uniprot` value should be unique (I think). Since that's the case, why not just make the `$uniref90` key in the uniref90 hash map to a set/list of `$uniprot` IDs. 

Once again, I'm using some python language here. Sorry if I'm over complicating/confusing the matter.

nilsoberg replied:
I can do that, but just keep it at hash refs for both for consistency.

rbdavid commented on lib/EFI/Sequence/Collection.pm:
```diff
@@ -0,0 +1,976 @@
+    if ($uniref90) {
+        $self->{uniref90}->{$uniref90}->{$uniprot} = 1;
+    }
+    $self->{uniprot}->{$uniprot} = [$uniref90, $uniref50];
```
Now this, this make sense to me.

rbdavid commented on lib/EFI/Sequence/Collection.pm:
```diff
@@ -0,0 +1,976 @@
+}
+
+
+# dangerous - should be called only from EFI::Sunburst::Data
```
Why is this dangerous?

nilsoberg replied:
It requires knowledge of the data structure that is returned.  It should be reworded because it's not dangerous, just not appropriate for other methods to call unless the format of the data is known by the caller.

rbdavid commented on lib/EFI/Sequence/Collection.pm:
```diff
@@ -0,0 +1,976 @@
+
+=head2 DESCRIPTION
+
+B<EFI::Sequence> is a Perl module used to represent a sequence from the EFI database
+with the sequence and attributes.
```
This is for __EFI::Sequence::Collection__ though. Copy paste issue?

nilsoberg replied:
Yes

rbdavid commented on lib/EFI/Sequence/Collection.pm:
```diff
@@ -0,0 +1,976 @@
+=head2 METHODS
+
+
+=head3 C<getFields()>
```
Currently, I think the order of subroutine documentation matches the order of subroutines defined in this file, but I feel like the documentation should start with the `load` and `save` subroutines and all associated subroutines called within. Then the get, set, add, etc subroutines can be presented after. Maybe a similar ordering of subroutine definitions should also be used to have the source code match the documentation?

nilsoberg replied:
I updated the POD order.  Leaving the source code the same so as to not create unnecessary changes in the history.

rbdavid commented on lib/EFI/Sequence/Collection.pm:
```diff
@@ -0,0 +1,976 @@
+=head3 C<addSequence($id, $attr, $seq)>
+
+Add a sequence to the collection.  Optionally add attributes (C<$attr> in the form of a hash ref)
+and a protein sequence C<$seq> as metadata.
```
Add some discussion about how redundant ID entries are handled.

rbdavid commented on lib/EFI/Sequence/Collection.pm:
```diff
@@ -0,0 +1,976 @@
+=item C<$seq> (optional)
+
+The protein amino acid sequence for the sequence.
+
```
Missing a "Returns" section about the retcode 0 or 1. Is this retcode ever used?

nilsoberg replied:
Return value indicates if the sequence was added or not (not added if already exists).

rbdavid commented on lib/EFI/Sequence/Collection.pm:
```diff
@@ -0,0 +1,976 @@
+This mapping will likely be a superset of the IDs added with the C<addSequence()> function in
+order to support sunburst diagrams for UniRef jobs (since all of the IDs are necessary, not
+just UniRef).
+
```
Could you add a discussion about the structure of the `$self->{uniref50}`, `$self->{uniref90}`, and `$self->{uniprot}` hashes?

nilsoberg replied:
I added a small discussion to the constructor and `removeSequence` which is where those hashes are most relevant.

rbdavid commented on lib/EFI/Sequence/Collection.pm:
```diff
@@ -0,0 +1,976 @@
+
+    my @ids = $seqs->getSequenceIds();
+    map { print "ID2 $_\n"; } @ids;
+
+
```
Missing documentation on the `getAllSequenceIds` subroutine.

nilsoberg replied:
I didn't document it because it shouldn't be a user-facing feature.  It is only called by `EFI::Sunburst::Data`.

rbdavid commented on lib/EFI/Sequence/Collection.pm:
```diff
@@ -0,0 +1,976 @@
+    map { print "ID2 $_\n"; } @ids;
+
+
+=head3 C<removeSequence($sequenceId)>
+
+Remove the sequence ID from the input metadata set if it is a primary sequence.  Also remove the
```
Why isn't this documentation neighboring the `addSequence` and `addUniref` subroutines' documentation?

nilsoberg replied:
Fixed

rbdavid commented on lib/EFI/Sequence/Collection.pm:
```diff
@@ -0,0 +1,976 @@
+UniRef IDs and the C<$sequenceId> is a UniRef ID, then all of the members of the UniRef cluster
+are also removed.  A few examples are given:
+
+B<Example 1: Input Originating in UniProt>
```
I think these example text/code blocks should explicitly define what `$self->{sequence_version}` is. Doing so may help highlight the different behaviors the subroutine will have.

nilsoberg replied:
Add an indicator to example title

rbdavid commented on lib/EFI/Sequence/Collection.pm:
```diff
@@ -0,0 +1,976 @@
+
+    # Initial metadata 
+    #UniProt_ID      Attribute       Value
+    #A0A8J3V1H9      Sequence_Source FAMILY
```
Should there be a line associated with the metadata for `A0A8J3TPF4`?

nilsoberg replied:
No, because the accession table has a mapping that may include more than what was in the input.

rbdavid commented on lib/EFI/Sequence/Collection.pm:
```diff
@@ -0,0 +1,976 @@
+
+    # Initial metadata 
+    #UniProt_ID      Attribute       Value
+    #A0A8J3V1H9      Sequence_Source FAMILY
```
As above, should there be a line associated with the metadata for `A0A8J3TPF4`?

nilsoberg replied:
No, because the accession table has a mapping that may include more than what was in the input.

rbdavid commented on lib/EFI/Sequence/Collection.pm:
```diff
@@ -0,0 +1,976 @@
+
+    # Initial metadata 
+    #UniProt_ID      Attribute       Value
+    #Q3AEU2	Sequence_Source FAMILY
```
Missing lines to mirror Example 3's "initial metadata" text?
```
#Q3AEU2     UniRef50_Cluster_Size   2
#Q3AEU2     UniRef50_IDs    A0A8J3TPF4^A0A8J3V1H9
```

nilsoberg replied:
Correct, fixed

rbdavid commented on lib/EFI/Sequence/Collection.pm:
```diff
@@ -0,0 +1,976 @@
+    #
+    # ID list after removal
+    #uniprot_id	uniref90_id	uniref50_id
+
```
Is there a missing example where the uniref90 Id of `A0A8J3V1H9` is removed? This would result in all information associated with both UniProt IDs being removed; an important additional demonstration of the subroutine's behavior.

nilsoberg replied:
Added an example.

rbdavid commented on lib/EFI/Sequence/Collection.pm:
```diff
@@ -0,0 +1,976 @@
+
+=head3 C<updateUnirefMetadata()>
+
+Creates or updates the UniRef-related metadata fields in the sequence metadata file.  For a
+UniRef90 sequence source these fields are C<UniRef90_Cluster_Size> which represents the number
+of UniProt IDs in the UniRef90 cluster, and C<UniRef50_IDs> which represents the IDs in the
+UniRef50 cluster.  (These are stored as a text string with each ID separated by the caret
+C<^> character.)  This information comes from the ID list.
```
I don't think this description explains the behavior well. Here's my try at it: 

"""
Creates or updates the UniRef-related metadata fields in the sequence metadata file. For a UniRef90 sequence source **,** these fields are "UniRef90_IDs" and "UniRef90_Cluster_Size". For a UniRef50 sequence source, these fields are "UniRef50_IDs" and "UniRef50_Cluster_Size". For both, the "Cluster_Size" field represents the number of UniProt Ids in the associated UniRef cluster. Similarly, the "IDs" field is a text string with each UniProt ID separated by the caret ``^`` character. This information comes from the ID list.
"""

nilsoberg replied:
Good rewrite. Updated.

rbdavid commented on lib/EFI/Sequence/Collection.pm:
```diff
@@ -0,0 +1,976 @@
+    $seqs->updateUnirefMetadata();
+
+
+
```
No documentation on the `loadMetadataFile`, `loadIdFile`, `saveMetadataFile`, nor `saveIdFile` subroutines.

nilsoberg replied:
There's no POD because they're not public functions.  Only called inside the module.

rbdavid commented on lib/EFI/Sequence/Collection.pm:
```diff
@@ -0,0 +1,976 @@
+    #A0A8J3V1H9	A0A8J3V1H9	Q3AEU2
+
+
+=head3 C<updateUnirefMetadata()>
```
Redundant docs header block for the `updateUnirefMetadata` subroutine.

nilsoberg replied:
Fixed

rbdavid commented on lib/EFI/Database/Schema.pm:
```diff
@@ -2,60 +2,174 @@
+
+=head2 NAME
+
+B<EFI::Database::Schema> - Perl module containing database schema constant
```
"constant" should be plural?

rbdavid commented on lib/EFI/Database/Schema.pm:
```diff
@@ -2,60 +2,174 @@
+
+=head4 Returns
+
+One of the constants listed below in L<CONSTANTS> (e.g. DBI_MYSQL).
```
Returns only one of the supported DBI constants? Or a list of all supported database formats?

nilsoberg replied:
Fixed

rbdavid commented on lib/EFI/Import/Util.pm:
```diff
@@ -0,0 +1,134 @@
+
+=head2 NAME
+
+B<EFI::Import::Util> - Perl module for B<EFI::Import> modules
```
I don't think this description is complete. Also, the the class/subroutines of this code are more focused on database querying rather than general import utilities, so I wonder if a better filename and module name is possible.

nilsoberg replied:
Created a separate issue to document the import process, issue #158.

rbdavid commented on lib/EFI/Import/Filter.pm:
```diff
@@ -4,147 +4,62 @@ package EFI::Import::Filter;
-    for (my $i = 0; $i < $max and $ids[$i]; $i++) {
-        $keep{$ids[$i]} = 1;
-    }
+package DummyStats;
```
Is this a placeholder for a more fleshed out package or a placeholder for instances where a stats package isn't necessary (so just use an empty class/module) in the `$self->{stats}` attribute?

nilsoberg replied:
It's the latter -- since subclasses expect the `$self->{stats}` to point to a valid object something needs to be inserted here if the filter isn't provided with a object for saving statistics about sequences that were filtered out.  The reason this is done is so that each filter can add custom statistics values.  There is probably a better of handling the stats but for now I think we should keep it this way to avoid doing more refactoring.

rbdavid commented on lib/EFI/Import/Filter.pm:
```diff
@@ -4,147 +4,62 @@ package EFI::Import::Filter;
     my $self = {};
+    $self->{dbh} = $args{dbh};
+    $self->{stats} = $args{stats} || DummyStats->new;
+    $self->{util} = new EFI::Import::Util(dbh => $self->{dbh});
```
`$self->{dbh}` and `$self->{util}` module attributes (I'm probably screwing up my language here) are getting redundant information about the database handle. Could the `$args{dbh}` information be accessed via `$self->{util}->{dbh}` instead of `$self->{dbh}`? This would halve the memory requirements associated with the database handle, even though that is likely very small.

rbdavid commented on pipelines/est/import/filter_ids.pl:
```diff
@@ -0,0 +1,100 @@
+
+use EFI::Database;
+use EFI::Import::Config::Filter;
+use EFI::Import::Filter;
```
Is this import line necessary since the module is never used but its children modules are?

nilsoberg replied:
No

rbdavid commented on lib/EFI/Import/Filter/Fraction.pm:
```diff
@@ -0,0 +1,41 @@
+    #TODO: keep SwissProts
+    my $numRemoved = 0;
+    for (my $i = 0; $i < @ids; $i++) {
+        $seqs->removeSequence($ids[$i]) and $numRemoved++ if ($i % $self->{fraction});
```
Given the same ordering of IDs, this will be deterministic. But, if the ordering changes (whether due to change in the EFI-DB version or some nuance of querying), recreating the same fractioning of the input IDs may not be recovered exactly. Should we add some text disclaimer about this on the frontend?

nilsoberg replied:
It's never been a concern.  You can raise it with Remi or John and depending on the response we can add this as a separate issue.

rbdavid commented on lib/EFI/Sequence/Type.pm:
```diff
@@ -0,0 +1,35 @@
+
+sub is_unknown_sequence {
+    my $seq = shift;
+    return $seq =~ m/^Z/i;
```
I'm not sure what this regular expression is capturing. Could you explain a bit? Is it just if the sequence string starts with "Z" or "z"?

rbdavid replied:
Or is it, if the "Z" character exists anywhere within the amino acid sequence? There are other single letter symbols used to denote sequence positions where the amino acid is unknown/ambiguous. In this context, "Z" just means either a glutamine or glutamic acid residue. "B" means either an asparagine or aspartic acid. "X" would be the symbol used to denote an unknown residue. 

The more I type this out, the less confident I am that this is what the regex is checking for.

nilsoberg replied:
If a sequence is included through the FASTA option and does not have a UniProt ID, then an ID is assigned to it in the "unknown" format (e.g. "`ZZZ###`").  This is not documented and I will add it.  Related to issue #36.

nilsoberg replied:
Those IDs always start with `Z`.

nilsoberg replied:
Added documentation.

rbdavid commented on lib/EFI/Import/Filter/Fragment.pm:
```diff
@@ -0,0 +1,45 @@
+    my $seqs = shift;
+
+    my @ids = $seqs->getAllSequenceIds();
+    @ids = grep { not is_unknown_sequence($_) } @ids;
```
If the sequence is unknown, then we won't know if its a fragment or not, so remove it. Is that the logic here?

rbdavid replied:
For example, if the user inputs a Query Sequence for Sequence BLAST input that does not map to a known UniProt ID, would it be labeled as "unknown" and then filtered out?

rbdavid replied:
Maybe I'm getting my order of operations backwards though.

nilsoberg replied:
Unknowns are all kept since the `grep` is there to obtain the IDs that will be filtered.  Therefore `@ids` only contains UniProt IDs and the fragment filtering is done on that set of IDs.  Only IDs in that array that are fragments are removed.

rbdavid commented on lib/EFI/Import/Filter/Fraction.pm:
Pod documentation would be helpful.

nilsoberg replied:
Created a separate issue to document the import process, issue #158.

rbdavid commented on lib/EFI/Import/Filter/Fragment.pm:
Pod documentation would be helpful.

nilsoberg replied:
Created a separate issue to document the import process, issue #158.

rbdavid commented on lib/EFI/Import/Filter/Taxonomy.pm:
Pod documentation would be immensely helpful, especially if users are able to upload their own taxonomy filtering files.

nilsoberg replied:
Created a separate issue to document the import process, issue #158.

rbdavid commented on pipelines/est/import/filter_ids.pl:
```diff
@@ -0,0 +1,100 @@
+
+# Restrict to specified families
+#TODO
+
```
Convert into an Issue?

nilsoberg replied:
See issues #29 and #153

rbdavid commented on lib/EFI/Import/Filter/Taxonomy.pm:
```diff
@@ -0,0 +1,128 @@
+    my $seqs = shift;
+
+    my @ids = $seqs->getAllSequenceIds();
+    @ids = grep { not is_unknown_sequence($_) } @ids;
```
See my comment in `EFI::Import::Filter::Fragment` about this.

nilsoberg replied:
My response is the same: this is only applying filtering to UniProt sequences and therefore unknown IDs are kept.

rbdavid commented on lib/EFI/Import/Filter/Taxonomy.pm:
```diff
@@ -0,0 +1,128 @@
+
+    my @conditions;
+    foreach my $cond (@{ $filter->{conditions} }) {
+        my $negate = (exists $cond->{negate} and $cond->{negate} eq "true");
+        my $useLike = (exists $cond->{exact} and $cond->{exact} eq "false");
+        my $compOp = $useLike ? ($negate ? "NOT LIKE" : "LIKE") : ($negate ? "!=" : "=");
```
Are there examples of "negate" and "exact" fields in a json format? Documentation and examples on this formatting would be hugely helpful. Maybe I just haven't found it yet.

nilsoberg replied:
The file format is not documented except in the "predefined" file in `assets/predefined_taxonomy_filters.yml`.  This documentation is necessary but in order to expedite the resolution of this PR I will convert this to an issue.

rbdavid commented on lib/EFI/Import/SequenceDB.pm:
```diff
@@ -9,7 +9,7 @@ use Capture::Tiny qw(capture);
 use Cwd qw(abs_path);
 use File::Basename qw(dirname);
-use lib dirname(abs_path(__FILE__)) . "/../../../../../../lib"; # Global libs
+use lib dirname(abs_path(__FILE__)) . "/../../lib"; # Global libs
```
Other module files don't include the `/lib` portion of this relative path. E.g. lib/EFI/Import/Filter/Taxonomy.pm has `use lib dirname(abs_path(__FILE__)) . "/../../../";`

nilsoberg replied:
Fixed.  `lib` should not be in any of the `use lib` directives.

rbdavid commented on lib/EFI/Import/SequenceDB.pm:
```diff
@@ -20,9 +20,9 @@ sub new {
 
     my $self = {};
     bless($self, $class);
```
Does `bless`ing before or after attributes get assigned to `$self` matter? In other module files, you `bless` right before `return $self`.

nilsoberg replied:
You need to `bless` before returning or before calling module functions but you can assign values to `$self` before or after `bless`.

rbdavid commented on lib/EFI/IdMapping.pm:
```diff
@@ -18,6 +18,7 @@ sub new {
     bless($self, $class);
```
Basic perl question: Does position of the `bless` command affect the object being created?

nilsoberg replied:
Yes, in that module (e.g. class) methods cannot be called on the object until after the `bless` has been called.

rbdavid commented on lib/EFI/Import/Statistics.pm:
```diff
@@ -14,8 +14,6 @@ sub new {
     my $self = {stats => {}};
     bless($self, $class);
 
-    $self->{mapping} = getMapping();
```
Removed the call to `getMapping()` and the attribute `$self->{mapping}`. But, kept the subroutine below. Maybe remove that code?

nilsoberg replied:
The web UI expects camel case fields, but the naming was inconsistent and needs to be rewritten to support JSON.  I decided for now to store the field names consistently (e.g. lower case, words separated by underscore) because not all of the attributes are read by the web UI.  I'm keeping `getMapping()` for historical reasons in case I decide to rename the JSON field names to be what the old web UI expects.

rbdavid commented on lib/EFI/EST/Metadata.pm:
```diff
@@ -1,244 +0,0 @@
-1;
-__END__
-
-=head1 EFI::Util::Metadata
```
Were any of these functions ported to other module files? Can we still use the documentation that you've got here?

nilsoberg replied:
The various attribute methods in **`EFI::Sequence`** and **`EFI::Sequence::Collection`** replaced this and were documented.

rbdavid commented on lib/EFI/Sunburst/Data.pm:
```diff
@@ -0,0 +1,176 @@
+}
+
+
+# sequenceIdData - EFI::Sequence::Collection
```
Do you mean `seqData` instead of `sequenceIdData`? 

More of a meta comment: I appreciate comments like this to inform what objects/modules are being passed into the subroutine. It would be described thoroughly in pod documentation but that text is far removed from the actual code, so having this basic info near the subroutine defintion is helpful, in my opinion.

nilsoberg replied:
Fixed.  I added documentation to this file.

rbdavid commented on lib/EFI/Sunburst/Data.pm:
Pod documentation would be helpful here.

nilsoberg replied:
Added

rbdavid commented on lib/EFI/Options.pm:
```diff
@@ -50,8 +50,8 @@ sub addOption {
     # $getoptName == test-arg=s
-    my $baseName = $getoptName =~ s/^(.+)=(.+?)$/$1/r;
-    my $optValType = $2;
+    my $baseName = $getoptName =~ s/^(.+)(=|:)(.+?)$/$1/r;
```
This change just enables either an equal or colon symbol to be used. Why is this necessary? 

Should the comment above be updated to highlight the possible strings that can be parsed?

nilsoberg replied:
This is simply a wrapper around Perl's **`Getopt::Long`**.  The `=` character means that an argument must have a value.  Using the colon `:` instead means that the value to an argument is optional.  I updated the comments as well as expanded the documentation.

rbdavid commented on lib/EFI/Import/Source.pm:
```diff
@@ -12,6 +12,8 @@ use lib dirname(abs_path(__FILE__)) . "/../../";
 use lib dirname(abs_path(__FILE__)) . "/../../../../../../lib"; # Global libs
```
Maybe I haven't been understanding these `use lib dirname(...)` lines. Isn't this path very far out of the scope of the EST repository?

nilsoberg replied:
This was a hold over from when these modules were stored in the `pipelines` folder hierarchy.  They were refactored so that all modules were contained in the `lib` directory and this long `use lib` was not removed.  Fixed.

rbdavid commented on lib/EFI/Options.pm:
```diff
@@ -362,9 +367,9 @@ Return information about the options that were added and parsed.
-with underscores C<_>.
+command line, even though it was added to the specification using C<addOption()>, it will
+not be present in this hash ref.  The option key is the option name provided in the
+specification to C<addOption> with the dash C<-> replaced with underscores C<_>.
```
So `getOptions` doesn't gather the complete mapping of all options that are being passed to the workflow? Like, if an option is left undefined on the command line, `getOptions` doesn't show that the option is still being used, just with its default value? If an option's default value were to change at some point in the future, how would users/designers know the option's value used for any past runs?

Not important for this PR, but maybe we should consider logging all options/parameters used on a run to some file for reproducibility sake.

nilsoberg replied:
This module doesn't do the same kind of validation that Python's `argparse` does.  It simply returns the intersection of the set of options that the user provided on the CLI and that were specified by the calls to `addOption()`.  For example, if the user provided `--flag --arg value --extra` on the command line, and only `--flag` and `--arg` were passed to `addOption()`, then the output from `getOptions()` is the hash ref `{flag => undef, arg => "value"}`.  If the user provided `--flag` on the CLI, and `--flag` and `--arg` were passed to `addOption()`, then the output from `getOptions()` is the hash ref `{flag => undef}`.  It is up to the script that uses **`EFI::Options`** to decide whether to continue or not.

However, this module does do validation in the sense that arg-values are correct (e.g. args that expect a value actually have a value), and verifies that all required arguments are present.  The caller must check the result of `parseOptions()` to see if the input arguments are valid.

I added this to the POD.

rbdavid commented on lib/EFI/Import/Config.pm:
```diff
@@ -4,213 +4,208 @@ package EFI::Import::Config;
+sub addImportOptions {
+    my $self = shift;
+    my %args = @_;
+    $self->addOption("output-dir=s", 0, "path to directory to store output in; if not specified, defaults to current working directory", OPT_DIR_PATH);
```
Is it possible to call, in this example `EFI::Options::OPT_DIR_PATH`? Constants/default values for all the EFI code are stored across many files. Being a bit more explicit would really help reviewers find the source of these constants.

rbdavid replied:
What does `use constant OPT_DIR_PATH => 5` mean in the context of a directory path?

nilsoberg replied:
**`EFI::Options`** does not follow best practices.  Perl has a concept of importing functions from external modules into the current namespace.  In Perl this is called "exporting" from one "package" into the current "package".  For example, in module `X`, using the `use Y qw(thing)` declaration tells Perl to make the `thing` function become part of the `X` namespace.  Perl can also implicitly import.  Best practices is for callers of a module to explicitly request the functions they want to import.  **`EFI::Options`** does not follow that practice in that the constants in `lib/EFI/Options.pm` are exported by default with the line `our @EXPORT = qw(OPT_VALUE OPT_FILE OPT_DIR_PATH OPT_PRINT_HELP OPT_ERRORS);`.  This means that every time code does a `use EFI::Options` the `OPT_VALUE`, `OPT_FILE`, `OPT_DIR_PATH`, `OPT_PRINT_HELP`, `OPT_ERRORS` constants implicitly become part of the namespace.  This is confusing.  While I generally follow best practices, this default export occurs in a few other places (`lib/EFI/Sequence/Type.pm` being one of them).  It should be changed, but not part of this PR.

(As a side note, Perl does not technically have constants.  Rather writing `use constant OPT_DIR_PATH => 5` results in a function called `OPT_DIR_PATH` that returns the value 5, in Perl `sub OPT_DIR_PATH { return 5; }`.)

In answer to your question about `OPT_DIR_PATH`, it is simply a flag that code that calls C<addOption()> uses to indicate that the value of the argument must be the path to a directory.

rbdavid commented on lib/EFI/Options.pm:
Could you add documentation about the constants and their values? Maybe I'm not fluent enough in perl, but setting `use constant OPT_ERRORS => 16;` doesn't make much sense to me.

nilsoberg replied:
I added documentation about the constants in `addOption()`.  However, `OPT_ERRORS` and `OPT_PRINT_HELP` were removed because they are no longer used.

rbdavid commented on lib/EFI/Import/Config.pm:
```diff
@@ -4,213 +4,208 @@ package EFI::Import::Config;
+    my $self = shift;
+    my %args = @_;
+    $self->addOption("output-dir=s", 0, "path to directory to store output in; if not specified, defaults to current working directory", OPT_DIR_PATH);
+    #TODO: document this
```
Yes please.

nilsoberg replied:
In the function documentation header.

rbdavid commented on lib/EFI/Import/Config.pm:
```diff
@@ -4,213 +4,208 @@ package EFI::Import::Config;
+    #TODO: document this
+    if ($args{include_config}) {
+        $self->addOption("efi-config-file=s", 1, "path to EFI database configuration file", OPT_FILE);
+        $self->addOption("efi-db=s", 1, "EFI database name, or path to EFI SQLite database file");
```
If this goes undefined, is that fatal? Should there be a default value for this? If I recall correctly, it can be defined in the EFI-DB config file?

nilsoberg replied:
It's up to the scripts to check the result of the parse/validation.  Currently both of these are required and all scripts verify that and die if it's not valid.  The `EFI::Import::Config*` modules all share the validation in this file so it always happens in this module so that the various import scripts don't need to duplicate code.

  In the future the database could be specified in the config file and only one file could be required.

rbdavid commented on lib/EFI/Import/Config.pm:
```diff
@@ -4,213 +4,208 @@ package EFI::Import::Config;
+#   ($status, $help) = $config->validateOptions();
+#   if ($help) {
+#       print "$help\n";
+#       exit(not $status); # if error, status is 0, so exit non zero to indicate to shell that there was a problem
```
Is there value to be gained by using a variety of return codes to signify which Import::Config child module is failing? Like, a return code of 100 would indicate that the `Import::Config::Sequences` `validateOptions` subroutine failed? 100 being just a random int that could be uniquely used to specify that one module/subroutine that is failing.

nilsoberg replied:
Possibly.  The way this is done is not optimal but since this collection of modules isn't user-facing I didn't spend a lot of time perfecting this.

rbdavid commented on lib/EFI/Import/Config.pm:
```diff
@@ -4,213 +4,208 @@ package EFI::Import::Config;
+B<EFI::Import::Config> is a utility module to get command line arguments for the EST import
+scripts.  The B<EFI::Import::Config::FastaImport>, B<EFI::Import::Config::Source>, and
+B<EFI::Import::Config::Sequences> modules derive from this and provide app-specific
+option parsing.  They should be used instead of directly using this module.
```
Is it worth noting that this module inherits from `EFI::Options`?

rbdavid replied:
Like, should "links" to the documentation for that module's subroutines be added to this and all inheriting modules' documentation? There's multiple layers of inheritance; `EFI::Import::Config` is just a middle man to the app-specific modules. For me, calls to ambiguous name spaces (e.g. `parseOptions()`) is making me search through many files to find what the subroutine does/returns.

So, like for `parseOptions()` the pod doc string could be:
```
=head3 C<parseOptions()>

Unmodified from B<EFI::Options>. See documentation in ../Options.pm, line AAA. 

=head4 Returns

{still list out what the subroutine returns}

```

Or something along those lines. Or, even just a list of inherited, unmodified subroutines would be helpful.

rbdavid replied:
I see now that you've done this for the `getOptions` subroutine, but I think you only do that since it has been slightly modified in this module.

nilsoberg replied:
I added some documentation that indicated that.  **`EFI::Import::Config::*` modules should really not inherit from **`EFI::Options`** because it obscures the proper function of the modules.  It should be refactored but outside of the scope of this PR.

rbdavid commented on lib/EFI/Import/Config/FastaImport.pm:
```diff
@@ -1,75 +1,58 @@
 
 
-sub validateAndProcessOptions {
+sub validateOptions {
```
Not much real validation of the input arguments happens here. 
- Should we check for whether the user-provided files (`uploaded-fasta`, `seq-mapping-file`, `seq-meta-file`) actually exist? If they have been declared to exist but don't, its not worth running through the rest of the import code to find out. 
- Is it worth checking that we can actually write a file to the `output-dir`? If we can't, its not worth running the rest of the import code to find that out.

nilsoberg replied:
First point: validation does occur because `uploaded-fasta` is a required argument, added in `addImportOptions()`.  Validation then occurs when calling `SUPER::validateOptions()` and checking the result.

Second point: other files all have default values that are implicitly defined if they are not explicitly provided.  The processes in the pipelines define these and automatically fail if the file was not written or doesn't exist.

Third point: If something like this happens, the script will die pretty quickly and then cause the Nextflow pipeline to terminate.  I don't think it's worth checking because the scripts are more or less internal to Nextflow and not user-facing.

rbdavid commented on pipelines/est/import/import_fasta.pl:
```diff
@@ -9,48 +9,74 @@
+                print "REMOVED $mappedId FROM import fasta\n";
+            }
+        }
     } elsif ($line =~ m/^>/) {
         $isValidSeq = 0;
```
This logic needs to be updated. I think the reassignment of `$isValidSeq = 0` needs to happen in the `else` block on line 71-74. 

Currently, this `elsif` block will never be executed since its the same regex check as the `if` block's. So, no sequences are ever invalidated based on their presence or absence in `$filteredSeqIds{$mappedId}`.

nilsoberg replied:
Fixed.

rbdavid commented on lib/EFI/Import/Config/Source.pm:
```diff
@@ -0,0 +1,85 @@
+    $self->SUPER::addImportOptions(include_config => 1);
+
+    $self->addOption("mode=s", 1, "the sequence retrieval mode (one of blast, family, accession, or fasta)");
+    $self->addOption("output-stats-file=s", 0, "output file to put sequence ID statistics into (defaults into --output-dir)", OPT_FILE);
```
Could these help strings also list the default file name that will be used?

nilsoberg replied:
They technically could.  However, the defaults are defined in `EFI::Import::Config::Defaults` so we'd either have to write them twice or get convoluted and add them here.  The module uses `use EFI::Import::Config::Defaults qw(get_default_path)` which I think is enough for a programmer to find the module and look at the constants.

rbdavid commented on lib/EFI/Import/Config/Source.pm:
```diff
@@ -0,0 +1,85 @@
+    $self->addOption("unmatched-ids=s", 0, "file containing IDs in FASTA or accession ID files that were not matched in the EFI database", OPT_FILE);
+    $self->addOption("blast-query=s", 0, "path to file containing sequence for initial BLAST; required for --mode blast", OPT_FILE);
+    $self->addOption("blast-output=s", 0, "output file to put BLAST results into; required for --mode blast", OPT_FILE);
+    #TODO: add domain support
+    #$self->addOption("domain-region=s", 0, "", "");
```
Issue for Domain feature to be added.

nilsoberg replied:
This will be added with issue #30

rbdavid commented on lib/EFI/Import/Config/Source.pm:
```diff
@@ -0,0 +1,85 @@
+    my @errors;
+
+    my $opts = $self->getOptions();
+    my $outputDir = $self->getOutputDir();
```
To validate that `$outputDir` will work, can you attempt to write a file to it? Better to check for failure here rather than failing to write the file later on in the import process.

nilsoberg replied:
Same as in a previous comment.  Since this is not user-facing we don't really care about trapping errors early on and displaying them to user.  I know I wrote the code and documentation as if it was user facing, but the pattern that I have generally followed is to die (e.g. throw exception) as soon as an error occurs (like being unable to write).  I prefer not to write new code to handle this case.  The user should know better...

rbdavid commented on lib/EFI/Import/Config/Source.pm:
```diff
@@ -0,0 +1,85 @@
+
+    $opts->{output_sunburst_ids_file} = get_default_path("sunburst_ids", $outputDir) if not $opts->{output_sunburst_ids_file};
+    $opts->{output_stats_file} = get_default_path("source_stats", $outputDir) if not $opts->{output_stats_file};
+
```
Can user-specified files be checked for their existence?

rbdavid replied:
Similar to what is done in `EFI::Import::Config:Sunburst`: 
`push @errors, "Error: invalid --sequence-meta-file path" if not -f $opts->{sequence_meta_file};`

nilsoberg replied:
Because these files are source-dependent, the existence is checked in the `EFI::Import::Source::*` modules.  For example, `--blast--output` is a required option, but only for `EFI::Import::Source::BLAST`, so when the `init()` function is called in that module, that parameter/option is checked for validity.  I think this applies to most of the places where you asked similar questions.

rbdavid commented on lib/EFI/Import/Config/Sequences.pm:
```diff
@@ -1,79 +1,62 @@
-    $self->addHelp("--fasta-db", "<BLAST_DB>", "Path to BLAST-formatted sequence database", 1);
-    $self->addHelp("--sequence-ids-file", "<ACCESSION_IDS_FILE>", "Path to text file containing list of accession IDs", 0);
-    $self->addHelp("--output-sequence-file", "<FASTA_FILE>", "Path to output file to put sequences in", 0);
+    $opts->{sequence_ids_file} = get_default_path("accession_table", $outputDir) if not $opts->{sequence_ids_file};
```
Do a similar `push @errors ...` check for file existence of the `sequence_ids_file`?

nilsoberg replied:
I got rid of this line because it shouldn't be implicitly defined anymore.  The code is shared between multiple pipelines so the `--sequence-ids-file` argument is now required (indicated by the `1` value when calling `addOption` for it).

rbdavid commented on lib/EFI/Import/Config/Sequences.pm:
```diff
@@ -1,79 +1,62 @@
 
-    push @$err, "Require --fasta-db" if not $h->{fasta_db};
+    my $opts = $self->getOptions();
+    my $outputDir = $self->getOutputDir();
```
Check to see if you can write to the `outputDir`?

nilsoberg replied:
See my response to your previous comments about this issue.  Also output dir is used for both read and write.  It's designed with Nextflow in mind since both input and output files could be located in the directory.

rbdavid commented on lib/EFI/Import/Config/Filter.pm:
```diff
@@ -0,0 +1,138 @@
+    my $extHelp = "Filter IDs to remove fragments, restrict to taxonomic categories, etc.";
+    my $self = $class->SUPER::new(%args, desc => $helpDesc, ext_desc => $extHelp);
+
+    die "Invalid predefined filter file '$args{predef_filter_file}' given" if $args{predef_filter_file} and not -f $args{predef_filter_file};
```
Wording could be improved. The file is not invalid, it doesn't exist! Also, should this `die` command happen in the `validateOptions` subroutine?

nilsoberg replied:
This is sort of a system level issue.  The user can provided a predef file on the command line, but there should always be a default predefined filter file if the user doesn't provided one, hence this check that occurs in the constructor.  ~~However, it would be better to move this check outside of the constructor to the caller (e.g. `pipelines/est/import/filter_ids.pl`) since that's where validation should really occur, especially since that module (Taxonomy) is the only place it is used.~~  I'm keeping it where it is because validation of that parameter has to occur there anyway.

rbdavid commented on lib/EFI/Import/Config/Filter.pm:
```diff
@@ -0,0 +1,138 @@
+    $self->addOption("sequence-ids-file=s", 0, "path to the output file to save filtered sequence IDs to (for sequence retrieval)", OPT_FILE);
+    $self->addOption("source-stats-file=s", 0, "path to the file containing source import stats", OPT_FILE);
+    $self->addOption("stats-file=s", 0, "path to the file to save filter statistics to (appends to source stats)", OPT_FILE);
+    #TODO: added in a future issue
+    #$self->addOption("restrict-family=s", 0, "", "");
+    #$self->addOption("restrict-domain=s", 0, "", "");
```
Add that issue.

nilsoberg replied:
Will be resolved by issues #30 and #153

rbdavid commented on lib/EFI/Import/Config/Filter.pm:
```diff
@@ -0,0 +1,138 @@
+    my $self = shift;
+    $self->SUPER::addImportOptions(include_config => 1);
+
+    $self->addOption("filter:s%", 0, "filters to apply (predef-name, predef-file, user-file, fragments, fraction)");
```
ahhh, this is where the colon delimiter is being used. Okay.

rbdavid commented on lib/EFI/Import/Config/Filter.pm:
```diff
@@ -0,0 +1,138 @@
+    my @errors;
+
+    my $opts = $self->getOptions();
+    my $outputDir = $self->getOutputDir();
```
Check to see if a file can be written to this outputDir before moving on.

nilsoberg replied:
See previous.  Let workflow handle errors.

rbdavid commented on lib/EFI/Import/Config/Filter.pm:
```diff
@@ -0,0 +1,138 @@
+
+    # Parse the filter options provided on the command line
+    my $filter = $opts->{filter} || {};
+    $opts->{fraction} = ($filter->{fraction} and $filter->{fraction} =~ m/^\d+$/) ? $filter->{fraction} : 1;
```
Today I learned about perl's contextual typing. I was thinking with my QA hat on, where the regex pattern would match "0000001" and I wondered how the code would handle it.

rbdavid replied:
What happens if the `$opts->{fraction}` value is larger than the total number of members in the family? Do we throw a warning message saying only one sequence was gathered from the set?

nilsoberg replied:
No warning is thrown.  Right now it is expected that the user knows what they're doing.

rbdavid commented on lib/EFI/Import/Source.pm:
```diff
@@ -30,19 +32,13 @@ sub new {
-    }
+    $self->{dbh} = $efiDbh || die "Require efi dbh argument";
+    $self->{sequence_version} = $config->{sequence_version} // SEQ_UNIPROT;
+    $self->{util} = new EFI::Import::Util(dbh => $self->{dbh});
```
Does `$self` need a `dbh` attribute if the `util` attribute also has a `dbh` attribute that can be accessed? Spinning up a `EFI::Import::Util` module requires the `$efiDbh` input already so why also save that as an attribute to this Source module?

nilsoberg replied:
Yes, it is use by modules that derive from this module (i.e. `EFI::Import::Source::*`) so it has to be cached.

rbdavid commented on lib/EFI/Import/Source.pm:
Add pod documentation would be helpful. What subroutines are to be inherited without modification, etc.

nilsoberg replied:
Created a separate issue to document the import process, issue #158.

rbdavid commented on lib/EFI/Import/Source/Accession.pm:
```diff
@@ -59,29 +59,36 @@ sub init {
 
 # 
-# getSequenceIds - called to obtain IDs from the accession ID file.  See parent class for usage.
+# loadFromSource - called to obtain IDs from the accession ID file.  See parent class for usage.
```
`EFI::Import::Source::loadFromSource` is an empty subroutine. What usage information is to be gathered from there?

nilsoberg replied:
Overridden in the source modules (e.g. BLAST, Accession, FASTA, Family).  Technically it doesn't need to be present in `Source.pm` but is given for clarity.  I added a comment to indicate this behavior.

rbdavid commented on lib/EFI/Import/Source/Accession.pm:
Pod documentation would be helpful.

nilsoberg replied:
Created a separate issue to document the import process, issue #158.

rbdavid commented on pipelines/est/import/get_sequence_ids.pl:
```diff
@@ -1,113 +1,111 @@
+
+
+# Access the primary source type (e.g. FASTA, Accession, BLAST, Family)
+#TODO: move the config parameter creation to this file (i.e. don't bury it in the Sources::XX modules)
```
Is this still the case?

nilsoberg replied:
You're right, this is no longer correct.  The `EFI::Import::Config::Source` module handles this.

rbdavid commented on pipelines/est/import/filter_ids.pl:
```diff
@@ -0,0 +1,100 @@
+    $taxFilter->applyFilter($seqData);
+}
+
+# Restrict to specified families
+#TODO
```
Add the issue before removing.

nilsoberg replied:
Done

nilsoberg replied:
Will be resolved by issue #153

rbdavid commented on pipelines/est/import/filter_ids.pl:
Major question/concern: Order of operations for applying filters will affect the results. Currently, the filters are applied in steps Fraction, Fragment, and then Taxonomy. Which sequences that pass the Fraction filter is a bit random and certainly uncaring to the biology; that filter just blindly removes based on a modulo boolean check. Whatever sequences remain may still not pass the other filters. This has the potential to really hamstring an analysis. 

Applying the Fraction filter last, would enable the other filters to thoroughly remove the biologically uninteresting sequences, leaving behind a set of sequences that we know the user is interested in, from which the Fraction filter can remove sequences in a more random, uncaring fashion. 

But, changing the order of filtering operations will/may return different results compared to those from the EFI v1 code base.

rbdavid replied:
Either way, I don't know if the order of operations is well documented on the website. Maybe it should be so that users can understand how their input mode's sequences map to the output sequences used to produce their SSNs.

nilsoberg replied:
Created a separate issue to document the import process, issue #158.

nilsoberg replied:
I would put this as a low priority question to ask John/Remi later this summer.  It can be added as a low priority issue if you want.

rbdavid replied:
Created as #161

rbdavid replied:
Issue #161 now made.

rbdavid commented on pipelines/est/import/get_sunburst_data.pl:
```diff
@@ -0,0 +1,77 @@
+
+
+
+sub saveToJson {
+    my $data = shift;
+    my $outputFile = shift;
+
+    $data = {
+        data => $data,
+    };
+
+    open my $fh, ">", $outputFile;
+
+    my $json = JSON->new->canonical(1);
+    if ($opts->{pretty_print}) {
+        $fh->print($json->pretty->encode($data));
+    } else {
+        $fh->print($json->encode($data));
+    }
+
+    close $fh;
+}
```
Should this be moved to `EFI::Sunburst::Data` module?

nilsoberg replied:
I'd prefer to keep it here since I was intending the `EFI::Sunburst::Data` module to be simply about retrieval.  I could imagine another code wanting the output at some point.  For example, we might want to generate the sunburst diagram ourselves.

rbdavid commented on pipelines/est/import/get_sunburst_data.pl:
```diff
@@ -0,0 +1,77 @@
+
+
+
+__END__
```
Is there missing Pod documentation? or just remove this.

nilsoberg replied:
Removed.  I added POD in some scripts but it's pretty much redundant because of the help added when using `EFI::Options` and I didn't add it here because of that.  The user could add `--help` and get the output.  The only reason to add POD to scripts is to create more documentation to explain complex features.

rbdavid commented on lib/EFI/Import/Config.pm:
```diff
@@ -4,213 +4,208 @@ package EFI::Import::Config;
-sub getOptions {
+sub getOutputDir {
     my $self = shift;
-    my $childOpts = shift || {};
-    my $childSpec = shift || [];
+    return $self->{opts}->{output_dir};
+}
```
Just for clarity/chronology of subroutine definitions, move this subroutine to below `addImportOptions`.

nilsoberg replied:
ok

rbdavid commented on lib/EFI/Import/Filter/Taxonomy.pm:
```diff
@@ -0,0 +1,128 @@
+
+
+1;
+__END__
```
Remove or keep and add docs.

rbdavid reviewed (commented):
Overall, very good. More documentation would be great but isn't highest priority at the moment. I left a whole bunch of questions and comments; most are low importance. I only noticed two major things:  

- Major issue: pipelines/est/import/import_fasta.pl, boolean checks in `if, elsif, else` block may result in no filtering actually being applied.
- Major question: pipelines/est/import/filter_ids.pl, order of applying the filters will affect the sequences that make it through the "import" step of the nextflow pipeline. Should that order of operations be changed?

Great work Nils!

Edit: Should have marked this as "Request Changes". Sorry!

nilsoberg commented:
Regarding your comment regarding placement of documentation:  I made a decision early on to put documentation for public functions in the POD at the end of the file, and write documentation at the start of a subroutine for internal functions (i.e. those called within the module only).  I think that embedding POD at the top of every function would make things cluttered (Perl's POD is very verbose).  It is not convenient to have to search the POD for the function documentation (although for us that use Vim using the `#` character and `:hsplit` work wonders).

nilsoberg committed:
Code and code comment cleanup

nilsoberg committed:
Add the invalid path to error messages on import scripts

nilsoberg committed:
Allow user-specified field separator for sequence metadata

nilsoberg committed:
Fix FASTA import issue

- import_fasta.pl was importing all sequences regardless of filtering
- Fix this by fixing logic

nilsoberg committed:
Remove unnecessary emits in workflow

nilsoberg committed:
Remove redundant code

nilsoberg committed:
Clarify UniRef ID association functions

nilsoberg committed:
Update and add POD

nilsoberg committed:
Add clarifying comments

nilsoberg committed:
Update help text for test environment script

nilsoberg committed:
Fix undefined reference issue added in previous commit

rbdavid commented on lib/EFI/Sequence/Collection.pm:
```diff
@@ -724,253 +753,143 @@ B<Example 4: Input Originating in UniRef50 (2)>
+a UniRef50 sequence source these fields are C<UniRef50_IDs> and C<UniRef50_Cluster_Size>.
+For both, the C<Cluster_Size> field represents the number of UniProt IDs in the associated
+UniRef cluster.  Similarly, the C<IDs> field is a text string with each UniProt ID separated
+the field separator character (defaults to caret C<^> but can be provided as a parameter to
```
You never close the open parentheses here.

rbdavid commented on lib/EFI/Sunburst/Data.pm:
```diff
@@ -173,4 +173,267 @@ sub traverseTree {
+
+=over
+
+=item C<$sequenceIdData>
```
This input argument is given the `$seqData` namespace not `$sequenceIdData` (line 25). Will also need to update line 228 to use `$seqData`.

nilsoberg replied:
Updated in POD, not in code.  Sometimes keep variable names shorter in code if they're used frequently.

rbdavid commented on lib/EFI/Annotations/Fields.pm:
```diff
@@ -41,10 +41,13 @@ use constant FIELD_COLOR_NODE_COUNT => "FIELD_COLOR_NODE_COUNT";
 
 use constant INPUT_SEQ_ID => "ZINPUTSEQ";
 
+use constant ANNO_ROW_SEP => "^";
```
I still think the default separator/delimiter should not be a sole carat character. Its likely rare to see in a annotation string but is it rare enough to not worry about?

nilsoberg replied:
Definitely rare enough not worry about it.  It's been this way for >7 years without an issue.  I believe it's only defined here so it should be easy to change.  Scripts can change it in the future when using the `EFI::Sequence::Collection` module.

rbdavid reviewed (approved):
Looks good to me. I think all the new issues referred to in the comments have been created.

rbdavid commented:
Just noticing, test 01d is failing before starting the nextflow run call. This is happening because the output directory associated with the test isn't created before the user-defined taxonomy filter file is created. Instead, lets include this user-provided taxonomy filter file in the suite of testing files instead of writing it during the test.

rbdavid commented on pipelines/est/est.nf:
```diff
@@ -1,53 +1,96 @@
+        path 'source_stats.json', emit: 'source_stats'
+        path 'blast_hits.tab', optional: true
+        path 'seq_mapping.tab', emit: 'seq_mapping', optional: true
+        path 'unmatched_id.tab', optional: true
```
Do we have a test that creates this `unmatched_id.tab` file?

rbdavid replied:
It is not saved in any of the results subdirectories so it looks like, no. I'll add it to an issue for implementation-test development.

nilsoberg replied:
The default test dataset doesn't have any unmatched IDs.  This file should be output from the fasta and accession configuration when IDs are included in the input but are not in the database.  Yes, it should be a test at some point in the future.

nilsoberg commented:
> Just noticing, test 01d is failing before starting the nextflow run call. This is happening because the output directory associated with the test isn't created before the user-defined taxonomy filter file is created. Instead, lets include this user-provided taxonomy filter file in the suite of testing files instead of writing it during the test.

The advantage of including the filter file in a temporary (e.g. test) directory is that we can write different test cases without having to add files to source control.  I will fix the code so that the file is written to the test directory using the test name.

nilsoberg committed:
Update rST documentation for Perl updates

nilsoberg committed:
Reorganize tests to distinguish between deployment and development tests

nilsoberg committed:
Merge branch 'nextflow-test' into 31-implement-family-addition-import
